### DISCUSSION TITLE
How to customize background color, font faces and get the page numbers in EPUB?

### DISCUSSION THREAD

**Discussion by nareshsvn** [readium/kotlin-toolkit#45] [Q&A]
In android  R2 kotlin sample app,
1. is there  a way we can apply Epub reader custom background colors (apart from readium  provided).
2. is there way we can apply custom font faces apart from readium provided font faces.
3. is there way we can get total number of pages and  current page number of Epub.
Many Thanks,

**Comment by mickael-menu** [comment]
> 1. is there  a way we can apply Epub reader custom background colors (apart from readium  provided).

There's nothing easily available for this yet. But you should be able to override the background color by calling `webView.setProperty("--USER__backgroundColor", "#ff0000")`. See `UserSettings.kt` for some examples on how `setProperty()` is called.

> 2. is there way we can apply custom font faces apart from readium provided font faces.

We plan to add this to the user settings but for now this is not possible without modifying the toolkit.

> 3. is there way we can get total number of pages and  current page number of Epub.

We have the notion of "position" which is equivalent to page numbers, but they don't match the exact screen pages because this is not easy to do in a reflowable reader.

You can get the total number of positions with:

```kotlin
publication.positions().size
```

And the current position is available when observing the current locator:

```kotlin
navigator.currentLocator.asLiveData().observe(activity, Observer { locator ->
    val currentPosition = locator?.locations?.position ?: 0
})
```

**Reply by nareshsvn** [reply]
**1. is there a way we can apply Epub reader custom background colors (apart from readium provided).**
    There's nothing easily available for this yet. But you should be able to override the background color by calling 
    webView.setProperty("--RS__backgroundColor", "#ff0000"). See UserSettings.kt for some examples on how setProperty() is called.

I tried using webView.setProperty("--RS__backgroundColor", "#ff0000") method. 
custom background color also reflecting but we are facing one problem like font color is not adopting accroding to the background color. 
For the custom background color, black color text only coming. is there a way we can make the reader text is adaptable according to the background color.(Ex: black background color and text should be in white color)

**Reply by mickael-menu** [reply]
> I tried using webView.setProperty("--RS__backgroundColor", "#ff0000") method.
> custom background color also reflecting but we are facing one problem like font color is not adopting accroding to the background color.
> For the custom background color, black color text only coming. is there a way we can make the reader text is adaptable according to the background color.(Ex: black background color and text should be in white color)

There's nothing to adjust the font color automatically, but there's also a property you can set for this: `--USER__textColor`. I also realized that maybe you should set `--USER__backgroundColor` instead of `--RS__backgroundColor`. See the documentation of Readium CSS for all the different properties available: https://github.com/readium/readium-css/blob/master/docs/CSS12-user_prefs.md#themes-background-and-text-colors

**Reply by nareshsvn** [reply]
Thank you. I am able to change the text color now. it looks like settings are applying only for the current chapter only.
but if I swipe few pages left/right(going to the next/previous chapter) side then the current background color and text colors are replaced by default settings like the white background and black color text.

**Reply by mickael-menu** [reply]
You have to make the changes in `UserSettings.kt` at the same place as `setProperty()` is called, otherwise they will be overwritten.
